### DISCUSSION TITLE
Non-linear fluorescence for simulation and trajectory reconstruction.

### DISCUSSION THREAD

**Discussion by b-grimaud** [DeepTrackAI/DeepTrack2#77] [Q&A] [answered]
Hello,

First and foremost, thank you for your work and for making it open source and available. The only alternative I could find for such a specific function was commercial.

I have a question regarding the `Fluorescence` module and noise simulation : My team is using two-photon microscopy to track nanoparticles. This means we have a relatively good SNR with little background noise, but second harmonic generation coupled with the very small size of our fluorophore means we have a very small amount of photons detected. 
This induces a lot of spatial noise, a 'jitter' of some sort.
Here's an example of a single static particle : 
![output-onlinegiftools](https://user-images.githubusercontent.com/83828302/138446154-463d8234-1517-47eb-a750-2593d4376118.gif)

Could there be a way to simulate such a phenomenon ?
For now I've only experimented simulation with regular, cohesive particles, and the quality of the predictions are limited by my simulations as well as the technical limitations of my computer and Google Colab.

Secondly, as I understand the prediction output is an array of coordinates were spots are located, for each frame. 
Have you considered implementing a 'linking' algorithm in order to link those spots into trajectories over time, frame by frame ?
We already use algorithmic approaches to do this, and I can reshape DeepTrack outputs to feed directly into it, but I would be interested to see if there are AI-based, parameters-free approches to this problem.

Thanks a lot !

**Comment by BenjaminMidtvedt** [comment, marked as answer]
Hi! We're happy to be of use!

I have definitely been able to get outputs resembling yours! Something like
```py
particle = dt.PointParticle()
optics = dt.Fluorescence(output_region=(0, 0, 64, 64))
pipeline = optics(particle) + dt.Poisson(snr=0.05)
pipeline.plot(cmap="gray")
```
![example](https://i.imgur.com/NSjWcLw.png)
gives something closely resembling your output! The snr in Poisson could more accurately be thought of as scaling with the photon count. The important thing is that the background is very close to 0.

Linking is something we are actively working on, and we expect to have a graph neural network-based particle tracer för linking frame-to-frame in a month or so!

**Reply by b-grimaud** [reply]
That's wonderful, thank you very much !

I will definetly keep track of the project for a linking model, we found that both published and homemade linking algorithms were very parameter-dependant, and we have to deliberately underperform to avoid false positives.

I had another question regarding generators, I am trying to better simulate my experimental data by larging some large artifcats that we often see.

My first instinct was to have a first dataset containing my "ground truth" particles, and another dataset containing the first dataset plus artifacts. Then, for training, I would feed the second dataset for images but only the first for labels.
I do not know, however, if both datasets will be "synchronized", as in the labels will match the image provided.

The generator I use is : 
```
if TRAIN_MODEL:
    generator = dt.generators.ContinuousGenerator(
        artifact_dataset,
        get_label(dataset.resolve()),
        batch_size=32,
        min_data_size=2e3,
        max_data_size=1e4
    )
```
I'm afraid that calling `resolve()` again in `get_label` generates another, separate dataset. Removing `get_label` raises an error about `'slice'` type objects being unashable.
While testing with an example provided in notebook n°4, 2.3 : 
```
for _ in range(NUMBER_OF_IMAGES):
    plt.figure(figsize=(15, 5))
    image_of_particle = dataset.resolve()
    artif_image = artifact_dataset.resolve()

    position = np.array(image_of_particle.get_property("position", get_one=False))
    plt.imshow(artif_image[..., 0], cmap="gray")
    plt.scatter(position[:, 1], position[:, 0],s=2)
    plt.show()
```
I understand that calling `resolve()` independantly on both datasets results in them being asynchronous, and I don't know if the same is applied within a generator.

**Reply by BenjaminMidtvedt** [reply]
Glad to help! 

The best approach depends a bit on your pipeline. In general, synchronization is not an issue, through there are some best practices to ensure that you don't fall into any edge-cases! Two features fall out of synchronization if `.update` is not called on both before they are resolved. In many cases, the two features in question may share most of the pipeline, so updating one will update the other. For example, your code may look something like:
```py
dataset = optics(particle) + noise ... # + whatever more you need
artifact_dataset = dataset + SomeFeatureThatAddsArtifacts
```

In this case, calling `artifact_dataset.update()` will update both artifact_dataset and dataset, so you can safely call `resolve()` on both of them independently, and they will stay in sync. Nonetheless, as you feared, your first example would not work, because the `.resolve()` is called once at initialization of the generator. However, there's a good chance that you don't need to change your generator at all! It is likely that:
```py
generator = dt.generators.ContinuousGenerator(
        artifact_dataset,
        get_label,
        batch_size=32,
        min_data_size=2e3,
        max_data_size=1e4
    )
```
will work just fine.

If your pipeline looks more like:
```py
dataset = optics(particle) + noise ... # + whatever more you need
artifact_dataset =  optics(particle) + noise + ... + SomeFeatureThatAddsArtifacts
```
then there is no way to keep them in sync, since they are to independent pipelines.

If you need to pass two features to the generator, one for data and one for labels, I recommend using dt.Combine. For example

```py
dataset = optics(particle) + noise ... # + whatever more you need
artifact_dataset = dataset + SomeFeatureThatAddsArtifacts

label = dataset + dt.Lambda(lambda: get_label)
sample_and_label = dt.Combine([artifact_dataset, label])

generator = dt.generators.ContinuousGenerator(
        sample_and_label,
        batch_size=32,
        min_data_size=2e3,
        max_data_size=1e4
    )
```
Here combine, well, combines two pipelines into one, which 100% ensures that they stay in sync. In this case you do not need even need to provide the generator with a `label_function`, because your feature already resolves both the image and the label.  

Hope this helps!

**Reply by b-grimaud** [reply]
Thanks again, this is exactly what I needed ! 

However, I encounter this error message while starting the generator you provided : 
```
Exception in thread Thread-7: before starting training
Traceback (most recent call last):
  File "/home/user/anaconda3/envs/dpt2/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/user/anaconda3/envs/dpt2/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user/anaconda3/envs/dpt2/lib/python3.8/site-packages/deeptrack/generators.py", line 324, in _continuous_get_training_data
    new_label = self.label_function(new_image)
TypeError: 'NoneType' object is not callable
```
I believe the `NoneType` error is due to the way I build the datasets.
As I understand it, the dataset provided in examples is a `features.Branch` object composed of a `optics.Microscope` object plus noise, etc.
If I add another `optics.Microscope` object after the first one and before noise, calling `resolve()` does return an `image.Image` object with the properties of both `Microscope` objects. However, simply displaying the `Image` with pyplot, as in the example I mentionned in my previous post, only shows a picture of particles resulting from the first `Microscope` object, with noise. I created both objects from indentical `Features` and through the same `Microscope`, to be sure, but the scatter still displays points that do not match visible particles.
![vis](https://user-images.githubusercontent.com/83828302/138895907-8bf71fac-a901-461f-84b3-2189f7151f76.png)
As the `position` property seems correctly updated (which I checked with a fixed position particle), and the documentation mentions `Features` of a `Branch` being called sequentially, it seems that my code only generates an image for the first `Microscope` object encountered.

Here's a reproducible example of the code : 
```
import deeptrack as dt
import matplotlib.pyplot as plt
import numpy as np

particle = dt.PointParticle(
    position=lambda: np.random.rand(2) * 512,
    position_unit="pixel",
    intensity=lambda: 100,)

art_particle = dt.PointParticle(
    position=lambda: np.random.rand(2) * 512,
    position_unit="pixel",
    intensity=lambda: 100,)

particles = particle**np.random.randint(50,80)
art_particles = art_particle**5

optics = dt.Fluorescence(
    NA=0.95,
    wavelength=520e-9,
    magnification=5,
    resolution=175e-9,
    output_region=(0, 0, 512, 512))

artifact_dataset = optics(particles) + optics(art_particles)

for _ in range(5):
    plt.figure(figsize=(15, 5))
    artifact_dataset.update()
    art_image = artifact_dataset.resolve()
    position = np.array(art_image.get_property("position", get_one=False))
    plt.imshow(art_image[..., 0], cmap="gray")
    plt.scatter(position[:, 1], position[:, 0],s=2)
    plt.show()
```
To get back to the code that you provided, I'm not sure which object type should `SomeFeatureThatAddsArtifacts` return, in order to for me to correctly add a `Feature` on top of another , and be able to call this code as you mentionned it :
```
label = dataset + dt.Lambda(lambda: get_label)
sample_and_label = dt.Combine([artifact_dataset, label])
```

I apologize if my questions seem trivial, as my team is composed of physicists and biologists we do not have anyone trained and dedicated for that kind of issue.

**Reply by BenjaminMidtvedt** [reply]
Completely my bad! I've been working on a development version where that was possible, forgot that it wasn't before!

I understand your confusion, since one expects `+` to things together. In the upcoming 2.1 release, your syntax would work flawlessly!

In your case, I would do something like this:
```
clean_dataset = optics(particles)
artifact_dataset = optics(art_particles)
dataset = dt.Combine([clean_dataset, artifact_dataset])

generator = dt.generators.ContinuousGenerator(
    dataset,
    batch_function=lambda tuple_clean_and_artifact: tuple_clean_and_artifact[0] + tuple_clean_and_artifact[1] # Add clean and artifact data
    label_function=lambda tuple_clean_and_artifact: get_label(tuple_clean_and_artifact[0])
   ... # rest of your arguments
)
```

To understand what happens, note that `dataset` now returns a tuple of two values, the clean data, and the added artifacts. In other words, you can do
```py
clean_image, artifacts = dataset.update().resolve()
```
to get the two images separately.

In generator, `batch_function` tells the generator what to do to go from the feature output (the tuple), to something it can put in a batch. You want to add them together in this case. `label_function` calculates the label, and since you only want positions from the clean image. you call `get_label` with the first value of the tuple, which is the clean image. 

Hopefully, this makes sense! I'm a biophysicist myself, so I completely understand.

**Reply by b-grimaud** [reply]
This works !

I can get both clean and "unclean" images, and get the positions from the first one only.

I did run into an issue during training, as the validation dataset is now defined as :
`validation_set = [dataset.update().resolve() for _ in range(validation_set_size)]` , with dataset returning a tuple, there is a dimension problem when the first epoch is completed and validation starts. 

I also tried something like this : 
```
validation_set = [dataset.update().resolve()[0] for _ in range(validation_set_size)]
art_validation_set = [dataset.update().resolve()[1] for _ in range(validation_set_size)]
validation_labels = [get_label(image) for image in validation_set]
```
and
```
validation_data=(
                              np.array(validation_set+art_validation_set),
                              np.array(validation_labels)
```
but the same error is raised : 
`ValueError: Input 0 of layer conv2d is incompatible with the layer: expected axis -1 of input shape to have value 1 but received input with shape (None, 2, 512, 512)`

For now, I've managed to make it run with only the clean dataset for validation, as a temporary measure.

I do not know if the training parameters (hyperparameters ?) provided in the example notebook n°4, that is `batch_size=32, min_data_size=2e3, max_data_size=1e4`, 5 epochs with weighted loss and 30 epochs with unweighted loss, is considered a minimum to get accurate results. I had to severely reduce those parameters due to RAM constraints both on my machine and Google Colab, with a weighted loss on the last epoch of : `loss: 0.0014 - nd_unet_crossentropy: 7.2263e-04 - val_loss: 0.0014 - val_nd_unet_crossentropy: 7.2297e-04`, and unweighted loss of : `loss: 7.2156e-04 - val_loss: 7.2297e-04`.
I have no point of reference for which order of magnitude is considered "good", but whenever I get predictions with those weights, using an image provided by the generator itself, the highest value in the resulting array is 1.25211315e-26.
Considering the threshold in the example is 0.99, I assume those results are abysmal. 

I don't know if it's because the smaller amount of data and epoch that I use for training, or the fact that the validation set does not include artifact.

**Reply by BenjaminMidtvedt** [reply]
Great! 

`np.array(validation_set+art_validation_set)` here `validation_set` and `art_validation_set` are lists, so it will do list concatenation. `np.array(validation_set)+np.array(art_validation_set)` should work better! If not, check the shape and make sure it matches (validation_set_size, 512, 512, 1). It can give a reference on where to debug.

It is completely fine to reduce the amount of data. In fact, I would assume `min_data_size`  and `max_data_size` of a few hundred is completely sufficient. However, if you reduce `min_data_size` by a factor 10, you need to increase the number of epochs by a factor 10 to make the training equivalent.  Also, a batch_size of 4-8 is enough to get a great model. If the model predicts essentially all zeros (as is your case), I would train it for longer using weighted loss, perhaps even increasing the weight of the weighted loss. I would for now skip the second stage where the model is trained on unweighted loss, and just do weighted loss. Make sure it is reasonable after this step. 

Training on unweighted loss is meant to further refine the model a bit to reduce false positives, but also makes the training slightly unstable, which is why it is not done from the start. I think your bad resultsstem from not training long enough on weighted loss to stabilize the model, since tha data size was reduced without increasing the number of epochs!

**Reply by b-grimaud** [reply]
Validation works !

I did manage to increase training time by increasing the number of epochs while keepint the batch size and dataset size relatively low.
However, one of the reason I initially reduced training time was because the loss, as well as the other metrics, wouldn't change after 2 or 3 epochs. I don't know if it's just a display issue or not, but prediction results were not proportionate to the amount of epochs used for training, even if the difference in training time is an order of magnitude.

For instance, training a model with 150 epochs results in this in the first few lines : 

```
Generating 200 / 200 samples before starting training
Epoch 1/150
50/50 [==============================] - 43s 509ms/step - loss: 0.0151 - nd_unet_crossentropy: 0.0774 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 2/150
50/50 [==============================] - 16s 328ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9165e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 3/150
50/50 [==============================] - 16s 327ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9218e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 4/150
50/50 [==============================] - 16s 325ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9165e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 5/150
50/50 [==============================] - 16s 326ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9165e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
```
Up to the last few : 

```
Epoch 148/150
50/50 [==============================] - 16s 329ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9192e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 149/150
50/50 [==============================] - 17s 340ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9192e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
Epoch 150/150
50/50 [==============================] - 17s 340ms/step - loss: 0.0015 - nd_unet_crossentropy: 7.9165e-04 - val_loss: 0.0015 - val_nd_unet_crossentropy: 7.9269e-04
```

As I didn't change anything in the model itself, I wonder if my data generator might be the issue here ? It seems like it's training on the same data over and over. Or maybe it has something to do with how Google Colab handles it, but that would seem strange.

**Reply by BenjaminMidtvedt** [reply]
To me, this looks like a model that predicts all 0s. You can validate that your generator is creating reasonable data using

```py
with generator:
    for data, label in generator:
           plt.subplot(1, 2, 1)
           plt.imshow(data[0, ..., 0])
           plt.subplot(1, 2, 2)
           plt.imshow(label[0, ..., 0])
           plt.show()
```

I would also look at if your images have reasonable intensities. If the max value is very low (1e-2 or lower), the model may not be able to train! In general, the pixel-intensity histogram of your training data should overlap nicely with the experimental data. You may also want to increase the loss weighting slightly, such as
```py
loss = dt.losses.flatten(
    dt.losses.weighted_crossentropy((25, 1))
)
```

As a point of reference, the weight number should be roughly (very roughly) 1 / np.mean(labels). So, if 0s are 10 times more common in your labels than 1s, it should be about 10. 

You'll probably be able to see if the model is learning at after only a few epochs. When things are working well, it learns very quickly!

**Reply by b-grimaud** [reply]
I found (one of) the culprit : I had deliberately lowered the `CIRCLE_RADIUS` parameter used to get labels to see if datasets would properly label ground truth particles and not artifactual ones. 

Correct me if I'm wrong, but as I understand it particle tracking models are essentially classifiers returning the probability of each pixel to be part of a valid spot, and subpixel accuracy is derived from the center of the label and not the spot itself, which would mean I need my label size to be as close as possible to the size of my experimental spots ?

Matching the size of the spots did improve training a lot, and I managed to get some results on generated data, albeit with a low threshold and some false positives and negatives.

The loss still stagnates around epoch 5 or 6, however. Increasing weight does lower the loss, and the rate at which it drops, but does not seem to extend the duration during which it decreases. Increasing the minimum amound of data also didn't change much.

> I would also look at if your images have reasonable intensities.

For now I use fixed, and fairly strong, intensities, as I am still learning how to use the `PointParticle` and `Fluorescence` objects, and playing around with the parameters to obtain something that is visually as close as possible to my experimental data, rather than sticking to the physical parameters that we use. Could that be an issue ?

Thanks !

Update : 

I messed around with several parameters, specifically minimum data size, batch size, weight, and even the size of the validation dataset. I got some pretty good results on simulated pictures, with no false positives and few false negatives in most cases. Results on experimental data are still rough, but I believe it could improve with some fine tuning of my data simulation model. 

However, the decrease in loss still stops pretty early. Pushing parameters lowers loss overall and improves the rate at which it decreases, but it is still stagnating after 10 or so epochs at most. I can essentially "brute force" the model to obtain a loss low enough to get good results before loss stagnates, but from what I've read and what you wrote before, it seems that proper trainining should take longer.

I also noticed issues when using `np.random.normal` instead of `rand` and `randint` when randomizing data, but that might just be an issue with TensorFlow itself !

**Reply by BenjaminMidtvedt** [reply]
Great! The rate of decrease and the time it takes for the training to stagnate depends mostly on the difficulty of the problem and is hard to predict from the start. I don't think 10 epochs is entirely unreasonable! Sometimes, decreasing the learning rate can help optimize the model further. In the end, I think you're on the home stretch! At this point it's mostly about testing hyper-parameters to optimize the results. 

Regardless, what will be orders of magnitude more important to the final quality is how well your simulation-pipeline approximates your experimental data. So that's where I would put most of my effort!
